File: tests/test_phase1.py
"""
Phase 1 Tests: Core Foundation
Tests for Component System, Entity System, and Game State
"""

import sys
import os
import json
import tempfile

# Add src to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from src.core import (
    # Components
    ComponentType,
    PositionComponent,
    HealthComponent,
    InventoryComponent,
    CurrencyComponent,
    RelationshipComponent,
    NeedsComponent,
    ScheduleComponent,
    OccupationComponent,
    # Entities
    Entity,
    EntityManager,
    create_player_entity,
    create_villager_entity,
    create_crop_entity,
    # Game State
    GameState,
    TimeOfDay,
    Season,
    Weather,
)


def test_component_creation():
    """Test 1.1: Create components and verify data storage"""
    print("\n=== Test 1.1: Component Creation ===")

    # Test Position Component
    pos = PositionComponent(location="VillageSquare", x=10.5, y=20.3)
    assert pos.location == "VillageSquare"
    assert pos.x == 10.5
    assert pos.y == 20.3
    print("✓ PositionComponent works")

    # Test Health Component
    health = HealthComponent(current=80, maximum=100)
    assert health.is_alive()
    health.damage(30)
    assert health.current == 50
    health.heal(20)
    assert health.current == 70
    print("✓ HealthComponent works")

    # Test Inventory Component
    inv = InventoryComponent(capacity=10)
    assert inv.add_item("wheat", 5)
    assert inv.has_item("wheat", 5)
    assert inv.get_count("wheat") == 5
    assert inv.remove_item("wheat", 2)
    assert inv.get_count("wheat") == 3
    print("✓ InventoryComponent works")

    # Test Currency Component
    currency = CurrencyComponent(amount=100)
    assert currency.has(50)
    assert currency.remove(30)
    assert currency.amount == 70
    currency.add(50)
    assert currency.amount == 120
    print("✓ CurrencyComponent works")

    # Test Relationship Component
    rel = RelationshipComponent()
    rel.set_relationship(entity_id=1, value=60)
    assert rel.get_relationship(1) == 60
    assert rel.get_relationship_level(1) == "friend"
    rel.modify_relationship(1, 20)
    assert rel.get_relationship(1) == 80
    assert rel.get_relationship_level(1) == "close_friend"
    print("✓ RelationshipComponent works")

    # Test Needs Component
    needs = NeedsComponent()
    needs.hunger = 30.0  # Low hunger (hungry)
    needs.energy = 20.0  # Low energy (tired)
    assert needs.get_most_urgent_need() == "energy"
    needs.eat(40)
    assert needs.hunger == 70.0
    print("✓ NeedsComponent works")

    # Test Schedule Component
    schedule = ScheduleComponent()
    schedule.set_activity("morning", "work")
    schedule.set_activity("evening", "socialize")
    assert schedule.get_activity("morning") == "work"
    assert schedule.get_activity("evening") == "socialize"
    print("✓ ScheduleComponent works")

    # Test Occupation Component
    occ = OccupationComponent(occupation="Blacksmith", workplace="Forge", skill_level=5)
    assert occ.occupation == "Blacksmith"
    assert occ.workplace == "Forge"
    assert occ.skill_level == 5
    print("✓ OccupationComponent works")

    print("\n✅ All component tests passed!")
    return True


def test_entity_system():
    """Test 1.2: Entity creation, component attachment/removal, entity lookup"""
    print("\n=== Test 1.2: Entity System ===")

    manager = EntityManager()

    # Create player entity
    player = create_player_entity(manager, "TestPlayer")
    assert player.name == "TestPlayer"
    assert player.entity_type == "Player"
    assert player.has_component(ComponentType.POSITION)
    assert player.has_component(ComponentType.HEALTH)
    assert player.has_component(ComponentType.INVENTORY)
    print(f"✓ Created player: {player}")

    # Create villager entity
    villager = create_villager_entity(manager, "Marcus", "Merchant")
    assert villager.name == "Marcus"
    assert villager.has_component(ComponentType.OCCUPATION)
    occ = villager.get_component(ComponentType.OCCUPATION)
    assert occ.occupation == "Merchant"
    print(f"✓ Created villager: {villager}")

    # Create crop entity
    crop = create_crop_entity(manager, "Wheat", "YourFarm", x=5.0, y=3.0)
    assert crop.name == "Wheat"
    assert crop.entity_type == "Crop"
    pos = crop.get_component(ComponentType.POSITION)
    assert pos.location == "YourFarm"
    assert pos.x == 5.0
    print(f"✓ Created crop: {crop}")

    # Test entity lookup
    assert manager.count_entities() == 3
    assert manager.get_entity(player.id) == player
    print(f"✓ Entity lookup works, total entities: {manager.count_entities()}")

    # Test query by component
    entities_with_position = manager.query_entities(ComponentType.POSITION)
    assert len(entities_with_position) == 3  # All have position
    print(f"✓ Query by component works: found {len(entities_with_position)} entities with position")

    entities_with_needs = manager.query_entities(ComponentType.NEEDS)
    assert len(entities_with_needs) == 1  # Only villager
    print(f"✓ Query found {len(entities_with_needs)} entities with needs (villager only)")

    # Test component removal
    villager.remove_component(ComponentType.SCHEDULE)
    assert not villager.has_component(ComponentType.SCHEDULE)
    print("✓ Component removal works")

    # Test entity removal
    manager.remove_entity(crop.id)
    assert manager.count_entities() == 2
    assert manager.get_entity(crop.id) is None
    print("✓ Entity removal works")

    print("\n✅ All entity system tests passed!")
    return True


def test_game_state():
    """Test 1.3: Game state creation, serialization, and loading"""
    print("\n=== Test 1.3: Game State System ===")

    # Create game state
    state = GameState()
    assert state.day_count == 1
    assert state.season == Season.SPRING
    assert state.time_of_day == TimeOfDay.MORNING
    print("✓ GameState initialized")

    # Create entities
    player = create_player_entity(state.entity_manager, "Hero")
    state.set_player(player.id)
    assert state.get_player() == player
    print(f"✓ Player set: {player.name}")

    villager1 = create_villager_entity(state.entity_manager, "Sarah", "Baker")
    villager2 = create_villager_entity(state.entity_manager, "Tom", "Farmer")
    print(f"✓ Created 2 villagers")

    # Modify some state
    inv = player.get_component(ComponentType.INVENTORY)
    inv.add_item("wheat", 10)
    inv.add_item("corn", 5)

    currency = player.get_component(ComponentType.CURRENCY)
    currency.add(250)

    # Set relationships
    rel = player.get_component(ComponentType.RELATIONSHIP)
    rel.set_relationship(villager1.id, 60)  # Friend
    rel.set_relationship(villager2.id, 30)  # Friendly
    print("✓ Modified player state")

    # Advance time
    initial_time = state.get_current_time_description()
    state.advance_time()
    assert state.time_of_day == TimeOfDay.AFTERNOON
    state.advance_time()
    state.advance_time()
    state.advance_time()  # Should wrap to next day
    assert state.day_count == 2
    assert state.time_of_day == TimeOfDay.MORNING
    print(f"✓ Time advancement works: {initial_time} -> {state.get_current_time_description()}")

    # Test statistics
    stats = state.get_statistics()
    print(f"✓ Statistics: {stats}")
    assert stats["total_entities"] == 3
    assert stats["villagers"] == 2

    # Test serialization
    print("\n--- Testing Serialization ---")
    state_dict = state.to_dict()
    assert "metadata" in state_dict
    assert "entities" in state_dict
    assert "time" in state_dict
    print("✓ to_dict() works")

    # Save to file
    with tempfile.NamedTemporaryFile(mode='w', suffix='.json', delete=False) as f:
        temp_file = f.name

    try:
        state.save_to_file(temp_file)
        print(f"✓ Saved to {temp_file}")

        # Load from file
        loaded_state = GameState.load_from_file(temp_file)
        print("✓ Loaded from file")

        # Verify loaded state
        assert loaded_state.day_count == state.day_count
        assert loaded_state.season == state.season
        assert loaded_state.entity_manager.count_entities() == 3
        print("✓ Loaded state matches original")

        # Verify player data
        loaded_player = loaded_state.get_player()
        assert loaded_player is not None
        assert loaded_player.name == "Hero"

        loaded_inv = loaded_player.get_component(ComponentType.INVENTORY)
        assert loaded_inv.get_count("wheat") == 10
        assert loaded_inv.get_count("corn") == 5

        loaded_currency = loaded_player.get_component(ComponentType.CURRENCY)
        assert loaded_currency.amount == 350  # 100 + 250

        loaded_rel = loaded_player.get_component(ComponentType.RELATIONSHIP)
        assert loaded_rel.get_relationship(villager1.id) == 60
        print("✓ Player data integrity verified")

        # Verify villagers
        villagers = loaded_state.entity_manager.get_entities_by_type("Villager")
        assert len(villagers) == 2
        baker = next(v for v in villagers if v.name == "Sarah")
        baker_occ = baker.get_component(ComponentType.OCCUPATION)
        assert baker_occ.occupation == "Baker"
        print("✓ Villager data integrity verified")

    finally:
        # Cleanup
        os.unlink(temp_file)

    print("\n✅ All game state tests passed!")
    return True


def run_all_tests():
    """Run all Phase 1 tests"""
    print("=" * 60)
    print("PHASE 1 TESTS: Core Foundation")
    print("Testing Component System, Entity System, and Game State")
    print("=" * 60)

    tests = [
        ("Component Creation", test_component_creation),
        ("Entity System", test_entity_system),
        ("Game State & Serialization", test_game_state),
    ]

    passed = 0
    failed = 0

    for test_name, test_func in tests:
        try:
            if test_func():
                passed += 1
        except Exception as e:
            print(f"\n❌ Test '{test_name}' failed with error:")
            print(f"   {type(e).__name__}: {e}")
            import traceback
            traceback.print_exc()
            failed += 1

    print("\n" + "=" * 60)
    print(f"RESULTS: {passed} passed, {failed} failed")
    print("=" * 60)

    if failed == 0:
        print("\n🎉 Phase 1: Core Foundation - COMPLETE! 🎉")
        print("\nSuccess Criteria Met:")
        print("✓ Can create entities with components")
        print("✓ Can maintain game state")
        print("✓ Can serialize to JSON")
        print("✓ Can load back from JSON")
        print("✓ Data integrity verified")
        return True
    else:
        print("\n⚠️  Some tests failed. Please review errors above.")
        return False


if __name__ == "__main__":
    success = run_all_tests()
    sys.exit(0 if success else 1)

File: tests/test_phase1.c
/**
 * test_phase1.c
 * Phase 1 Tests: Core Foundation
 * Tests for Component System, Entity System, and Game State
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "../lib/cJSON.h"
#include "../src/core/component.h"
#include "../src/core/entity.h"
#include "../src/core/game_state.h"

// Test counter
static int tests_passed = 0;
static int tests_failed = 0;

#define TEST_ASSERT(condition, message) \
    do { \
        if (!(condition)) { \
            printf("  ✗ FAILED: %s\n", message); \
            tests_failed++; \
            return false; \
        } \
    } while(0)

#define TEST_PASS(message) \
    do { \
        printf("  ✓ %s\n", message); \
        tests_passed++; \
    } while(0)

// ============================================================================
// Component Tests
// ============================================================================

bool test_component_creation() {
    printf("\n=== Test 1.1: Component Creation ===\n");

    // Test Position Component
    PositionComponent* pos = position_component_create("VillageSquare", 10.5f, 20.3f);
    TEST_ASSERT(pos != NULL, "Position component created");
    TEST_ASSERT(strcmp(pos->location, "VillageSquare") == 0, "Position location correct");
    TEST_ASSERT(pos->x == 10.5f && pos->y == 20.3f, "Position coordinates correct");
    TEST_PASS("PositionComponent works");
    position_component_destroy(pos);

    // Test Health Component
    HealthComponent* health = health_component_create(80, 100);
    TEST_ASSERT(health != NULL, "Health component created");
    TEST_ASSERT(health_component_is_alive(health), "Health component is alive");
    health_component_damage(health, 30);
    TEST_ASSERT(health->current == 50, "Health damage works");
    health_component_heal(health, 20);
    TEST_ASSERT(health->current == 70, "Health heal works");
    TEST_PASS("HealthComponent works");
    health_component_destroy(health);

    // Test Inventory Component
    InventoryComponent* inv = inventory_component_create(10);
    TEST_ASSERT(inv != NULL, "Inventory component created");
    TEST_ASSERT(inventory_component_add_item(inv, "wheat", 5), "Add item works");
    TEST_ASSERT(inventory_component_has_item(inv, "wheat", 5), "Has item works");
    TEST_ASSERT(inventory_component_get_count(inv, "wheat") == 5, "Get count works");
    TEST_ASSERT(inventory_component_remove_item(inv, "wheat", 2), "Remove item works");
    TEST_ASSERT(inventory_component_get_count(inv, "wheat") == 3, "Item count correct after removal");
    TEST_PASS("InventoryComponent works");
    inventory_component_destroy(inv);

    // Test Currency Component
    CurrencyComponent* currency = currency_component_create(100);
    TEST_ASSERT(currency != NULL, "Currency component created");
    TEST_ASSERT(currency_component_has(currency, 50), "Has currency works");
    TEST_ASSERT(currency_component_remove(currency, 30), "Remove currency works");
    TEST_ASSERT(currency->amount == 70, "Currency amount correct");
    currency_component_add(currency, 50);
    TEST_ASSERT(currency->amount == 120, "Add currency works");
    TEST_PASS("CurrencyComponent works");
    currency_component_destroy(currency);

    // Test Relationship Component
    RelationshipComponent* rel = relationship_component_create();
    TEST_ASSERT(rel != NULL, "Relationship component created");
    relationship_component_set(rel, 1, 60);
    TEST_ASSERT(relationship_component_get(rel, 1) == 60, "Set relationship works");
    TEST_ASSERT(strcmp(relationship_component_get_level(rel, 1), "friend") == 0, "Relationship level correct");
    relationship_component_modify(rel, 1, 20);
    TEST_ASSERT(relationship_component_get(rel, 1) == 80, "Modify relationship works");
    TEST_ASSERT(strcmp(relationship_component_get_level(rel, 1), "close_friend") == 0, "Relationship level updated");
    TEST_PASS("RelationshipComponent works");
    relationship_component_destroy(rel);

    // Test Needs Component
    NeedsComponent* needs = needs_component_create();
    TEST_ASSERT(needs != NULL, "Needs component created");
    needs->hunger = 30.0f;  // Low hunger (hungry)
    needs->energy = 20.0f;  // Low energy (tired)
    TEST_ASSERT(strcmp(needs_component_get_most_urgent(needs), "energy") == 0, "Most urgent need correct");
    needs_component_eat(needs, 40.0f);
    TEST_ASSERT(needs->hunger == 70.0f, "Eat works");
    TEST_PASS("NeedsComponent works");
    needs_component_destroy(needs);

    // Test Schedule Component
    ScheduleComponent* schedule = schedule_component_create();
    TEST_ASSERT(schedule != NULL, "Schedule component created");
    schedule_component_set_activity(schedule, "morning", "work");
    schedule_component_set_activity(schedule, "evening", "socialize");
    TEST_ASSERT(strcmp(schedule_component_get_activity(schedule, "morning"), "work") == 0, "Schedule get works");
    TEST_ASSERT(strcmp(schedule_component_get_activity(schedule, "evening"), "socialize") == 0, "Schedule set works");
    TEST_PASS("ScheduleComponent works");
    schedule_component_destroy(schedule);

    // Test Occupation Component
    OccupationComponent* occ = occupation_component_create("Blacksmith", "Forge", 5);
    TEST_ASSERT(occ != NULL, "Occupation component created");
    TEST_ASSERT(strcmp(occ->occupation, "Blacksmith") == 0, "Occupation correct");
    TEST_ASSERT(strcmp(occ->workplace, "Forge") == 0, "Workplace correct");
    TEST_ASSERT(occ->skill_level == 5, "Skill level correct");
    TEST_PASS("OccupationComponent works");
    occupation_component_destroy(occ);

    printf("\n✅ All component tests passed!\n");
    return true;
}

// ============================================================================
// Entity System Tests
// ============================================================================

bool test_entity_system() {
    printf("\n=== Test 1.2: Entity System ===\n");

    EntityManager* manager = entity_manager_create();
    TEST_ASSERT(manager != NULL, "EntityManager created");

    // Create player entity
    Entity* player = create_player_entity(manager, "TestPlayer");
    TEST_ASSERT(player != NULL, "Player entity created");
    TEST_ASSERT(strcmp(player->name, "TestPlayer") == 0, "Player name correct");
    TEST_ASSERT(strcmp(player->entity_type, "Player") == 0, "Player type correct");
    TEST_ASSERT(entity_has_component(player, COMPONENT_POSITION), "Player has position");
    TEST_ASSERT(entity_has_component(player, COMPONENT_HEALTH), "Player has health");
    TEST_ASSERT(entity_has_component(player, COMPONENT_INVENTORY), "Player has inventory");
    printf("  ✓ Created player: Entity(%d, '%s', type=%s, components=%d)\n",
           player->id, player->name, player->entity_type, player->component_count);

    // Create villager entity
    Entity* villager = create_villager_entity(manager, "Marcus", "Merchant", "VillageSquare");
    TEST_ASSERT(villager != NULL, "Villager entity created");
    TEST_ASSERT(strcmp(villager->name, "Marcus") == 0, "Villager name correct");
    TEST_ASSERT(entity_has_component(villager, COMPONENT_OCCUPATION), "Villager has occupation");

    OccupationComponent* occ = (OccupationComponent*)entity_get_component(villager, COMPONENT_OCCUPATION);
    TEST_ASSERT(occ != NULL, "Get occupation component works");
    TEST_ASSERT(strcmp(occ->occupation, "Merchant") == 0, "Occupation correct");
    printf("  ✓ Created villager: Entity(%d, '%s', type=%s, components=%d)\n",
           villager->id, villager->name, villager->entity_type, villager->component_count);

    // Create crop entity
    Entity* crop = create_crop_entity(manager, "Wheat", "YourFarm", 5.0f, 3.0f);
    TEST_ASSERT(crop != NULL, "Crop entity created");
    TEST_ASSERT(strcmp(crop->name, "Wheat") == 0, "Crop name correct");
    TEST_ASSERT(strcmp(crop->entity_type, "Crop") == 0, "Crop type correct");

    PositionComponent* pos = (PositionComponent*)entity_get_component(crop, COMPONENT_POSITION);
    TEST_ASSERT(pos != NULL, "Crop has position");
    TEST_ASSERT(strcmp(pos->location, "YourFarm") == 0, "Crop location correct");
    TEST_ASSERT(pos->x == 5.0f, "Crop x position correct");
    printf("  ✓ Created crop: Entity(%d, '%s', type=%s, components=%d)\n",
           crop->id, crop->name, crop->entity_type, crop->component_count);

    // Test entity lookup
    TEST_ASSERT(entity_manager_count(manager) == 3, "Entity count correct");
    TEST_ASSERT(entity_manager_get_entity(manager, player->id) == player, "Get entity by ID works");
    printf("  ✓ Entity lookup works, total entities: %d\n", entity_manager_count(manager));

    // Test query by component
    Entity* entities_with_position[MAX_ENTITIES];
    ComponentType position_type = COMPONENT_POSITION;
    int count = entity_manager_query_entities(manager, &position_type, 1, entities_with_position, MAX_ENTITIES);
    TEST_ASSERT(count == 3, "Query by component works");
    printf("  ✓ Query by component works: found %d entities with position\n", count);

    Entity* entities_with_needs[MAX_ENTITIES];
    ComponentType needs_type = COMPONENT_NEEDS;
    count = entity_manager_query_entities(manager, &needs_type, 1, entities_with_needs, MAX_ENTITIES);
    TEST_ASSERT(count == 1, "Query finds only entities with needs");
    printf("  ✓ Query found %d entities with needs (villager only)\n", count);

    // Test entity removal
    TEST_ASSERT(entity_manager_remove_entity(manager, crop->id), "Remove entity works");
    TEST_ASSERT(entity_manager_count(manager) == 2, "Entity count after removal correct");
    TEST_ASSERT(entity_manager_get_entity(manager, crop->id) == NULL, "Removed entity not found");
    TEST_PASS("Entity removal works");

    entity_manager_destroy(manager);

    printf("\n✅ All entity system tests passed!\n");
    return true;
}

// ============================================================================
// Game State Tests
// ============================================================================

bool test_game_state() {
    printf("\n=== Test 1.3: Game State System ===\n");

    // Create game state
    GameState* state = game_state_create();
    TEST_ASSERT(state != NULL, "GameState created");
    TEST_ASSERT(state->day_count == 1, "Initial day count correct");
    TEST_ASSERT(state->season == SEASON_SPRING, "Initial season correct");
    TEST_ASSERT(state->time_of_day == TIME_MORNING, "Initial time of day correct");
    TEST_PASS("GameState initialized");

    // Create entities
    Entity* player = create_player_entity(state->entity_manager, "Hero");
    TEST_ASSERT(player != NULL, "Player created");
    game_state_set_player(state, player->id);
    TEST_ASSERT(game_state_get_player(state) == player, "Set/get player works");
    printf("  ✓ Player set: %s\n", player->name);

    Entity* villager1 = create_villager_entity(state->entity_manager, "Sarah", "Baker", "VillageSquare");
    Entity* villager2 = create_villager_entity(state->entity_manager, "Tom", "Farmer", "VillageSquare");
    TEST_ASSERT(villager1 != NULL && villager2 != NULL, "Villagers created");
    printf("  ✓ Created 2 villagers\n");

    // Modify player state
    InventoryComponent* inv = (InventoryComponent*)entity_get_component(player, COMPONENT_INVENTORY);
    TEST_ASSERT(inv != NULL, "Player inventory exists");
    inventory_component_add_item(inv, "wheat", 10);
    inventory_component_add_item(inv, "corn", 5);

    CurrencyComponent* currency = (CurrencyComponent*)entity_get_component(player, COMPONENT_CURRENCY);
    TEST_ASSERT(currency != NULL, "Player currency exists");
    currency_component_add(currency, 250);

    // Set relationships
    RelationshipComponent* rel = (RelationshipComponent*)entity_get_component(player, COMPONENT_RELATIONSHIP);
    TEST_ASSERT(rel != NULL, "Player relationships exist");
    relationship_component_set(rel, villager1->id, 60);
    relationship_component_set(rel, villager2->id, 30);
    TEST_PASS("Modified player state");

    // Test time advancement
    const char* initial_time = game_state_get_time_description(state);
    printf("  ✓ Time: %s\n", initial_time);

    game_state_advance_time(state);
    TEST_ASSERT(state->time_of_day == TIME_AFTERNOON, "Time advanced to afternoon");

    game_state_advance_time(state);
    game_state_advance_time(state);
    game_state_advance_time(state);  // Should wrap to next day
    TEST_ASSERT(state->day_count == 2, "Day count advanced");
    TEST_ASSERT(state->time_of_day == TIME_MORNING, "Time wrapped to morning");

    printf("  ✓ Time advancement works: %s -> %s\n", initial_time, game_state_get_time_description(state));

    // Test serialization
    printf("\n--- Testing Serialization ---\n");
    cJSON* state_json = game_state_to_json(state);
    TEST_ASSERT(state_json != NULL, "to_json works");
    TEST_PASS("Serialization to JSON works");

    // Save to file
    const char* temp_file = "/tmp/test_game_save.json";
    TEST_ASSERT(game_state_save_to_file(state, temp_file), "Save to file works");

    // Load from file
    GameState* loaded_state = game_state_load_from_file(temp_file);
    TEST_ASSERT(loaded_state != NULL, "Load from file works");
    TEST_PASS("Loaded from file");

    // Verify loaded state
    TEST_ASSERT(loaded_state->day_count == state->day_count, "Day count preserved");
    TEST_ASSERT(loaded_state->season == state->season, "Season preserved");
    TEST_ASSERT(entity_manager_count(loaded_state->entity_manager) == 3, "Entity count preserved");
    TEST_PASS("Loaded state matches original");

    // Verify player data
    Entity* loaded_player = game_state_get_player(loaded_state);
    TEST_ASSERT(loaded_player != NULL, "Player loaded");
    TEST_ASSERT(strcmp(loaded_player->name, "Hero") == 0, "Player name preserved");

    InventoryComponent* loaded_inv = (InventoryComponent*)entity_get_component(loaded_player, COMPONENT_INVENTORY);
    TEST_ASSERT(loaded_inv != NULL, "Player inventory loaded");
    TEST_ASSERT(inventory_component_get_count(loaded_inv, "wheat") == 10, "Wheat count preserved");
    TEST_ASSERT(inventory_component_get_count(loaded_inv, "corn") == 5, "Corn count preserved");

    CurrencyComponent* loaded_currency = (CurrencyComponent*)entity_get_component(loaded_player, COMPONENT_CURRENCY);
    TEST_ASSERT(loaded_currency != NULL, "Player currency loaded");
    TEST_ASSERT(loaded_currency->amount == 350, "Currency amount preserved (100 + 250)");

    RelationshipComponent* loaded_rel = (RelationshipComponent*)entity_get_component(loaded_player, COMPONENT_RELATIONSHIP);
    TEST_ASSERT(loaded_rel != NULL, "Player relationships loaded");
    TEST_ASSERT(relationship_component_get(loaded_rel, villager1->id) == 60, "Relationship with Sarah preserved");
    TEST_PASS("Player data integrity verified");

    // Verify villagers
    Entity* villagers[MAX_ENTITIES];
    int villager_count = entity_manager_get_entities_by_type(loaded_state->entity_manager, "Villager", villagers, MAX_ENTITIES);
    TEST_ASSERT(villager_count == 2, "Villager count correct");

    Entity* baker = NULL;
    for (int i = 0; i < villager_count; i++) {
        if (strcmp(villagers[i]->name, "Sarah") == 0) {
            baker = villagers[i];
            break;
        }
    }
    TEST_ASSERT(baker != NULL, "Baker villager found");

    OccupationComponent* baker_occ = (OccupationComponent*)entity_get_component(baker, COMPONENT_OCCUPATION);
    TEST_ASSERT(baker_occ != NULL, "Baker occupation loaded");
    TEST_ASSERT(strcmp(baker_occ->occupation, "Baker") == 0, "Baker occupation preserved");
    TEST_PASS("Villager data integrity verified");

    // Cleanup
    cJSON_Delete(state_json);
    game_state_destroy(state);
    game_state_destroy(loaded_state);

    printf("\n✅ All game state tests passed!\n");
    return true;
}

// ============================================================================
// Main Test Runner
// ============================================================================

int main() {
    printf("============================================================\n");
    printf("PHASE 1 TESTS: Core Foundation\n");
    printf("Testing Component System, Entity System, and Game State\n");
    printf("============================================================\n");

    bool all_passed = true;

    // Run tests
    if (!test_component_creation()) all_passed = false;
    if (!test_entity_system()) all_passed = false;
    if (!test_game_state()) all_passed = false;

    // Print results
    printf("\n============================================================\n");
    printf("RESULTS: %d passed, %d failed\n", tests_passed, tests_failed);
    printf("============================================================\n");

    if (all_passed && tests_failed == 0) {
        printf("\n🎉 Phase 1: Core Foundation - COMPLETE! 🎉\n");
        printf("\nSuccess Criteria Met:\n");
        printf("✓ Can create entities with components\n");
        printf("✓ Can maintain game state\n");
        printf("✓ Can serialize to JSON\n");
        printf("✓ Can load back from JSON\n");
        printf("✓ Data integrity verified\n");
        return 0;
    } else {
        printf("\n⚠️  Some tests failed. Please review errors above.\n");
        return 1;
    }
}
